src/main/java/com/takipi/oss/storage/resources/fs/multifetcher/ConcurrentMultiFetcher.java

```diff
@@ -0,0 +1,108 @@
+package com.takipi.oss.storage.resources.fs.multifetcher;
+
+import com.takipi.oss.storage.data.EncodingType;
+import com.takipi.oss.storage.data.RecordWithData;
+import com.takipi.oss.storage.data.fetch.MultiFetchRequest;
+import com.takipi.oss.storage.data.fetch.MultiFetchResponse;
+import com.takipi.oss.storage.fs.Record;
+import com.takipi.oss.storage.fs.api.Filesystem;
+import com.takipi.oss.storage.fs.cache.Cache;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class ConcurrentMultiFetcher extends BaseMultiFetcher {
+	
+	private static final Logger logger = LoggerFactory.getLogger(ConcurrentMultiFetcher.class);
+	
+	private final ExecutorService executorService;
+	private final AtomicInteger threadCount = new AtomicInteger();
+	
+	public ConcurrentMultiFetcher(int maxThreads) {
+		
+		if (maxThreads > 50) {
+			logger.warn("ConcurrentMultiFetcher concurrency level can not be greater than 50");
+			maxThreads = 50;
+		}
+		else if (maxThreads < 2) {
+			logger.warn("ConcurrentMultiFetcher concurrency level can not be less than 2");
+			maxThreads = 2;
+		}
+		
+		logger.info("ConcurrentMultiFetcher maximum number of threads = " + maxThreads);
+		
+		ThreadFactory threadFactory = new ThreadFactory() {
+			@Override
+			public Thread newThread(Runnable r)
+			{
+				Thread t = new Thread(r);
+				t.setDaemon(true);
+				t.setName("fetcher_thread_" + threadCount.incrementAndGet());
+				return t;
+			}
+		};
+		
+		executorService = Executors.newFixedThreadPool(maxThreads, threadFactory);
+	}
+	
+	@Override
+	public MultiFetchResponse loadData(final MultiFetchRequest request, final Filesystem<Record> filesystem) {
+		
+		final EncodingType encodingType = request.encodingType;
+		final List<Record> recordsToRetrieve = request.records;
+		final int count = recordsToRetrieve.size();
+		final List<Future<String>> futures = new ArrayList<>(count);
```

Comment:
Lists.newArrayListWIthCapacity(...)

Reply:
I left it as new ArrayList because newArrayListWIthCapacity merely added a check for a negative capacity which in my usages is certain to be >= 0